File: newton/newton.js
/*

    _               ___         _      _
 _ | |__ ___ ____ _/ __| __ _ _(_)_ __| |_
| || / _` \ V / _` \__ \/ _| '_| | '_ \  _|
 \__/\__,_|\_/\__,_|___/\__|_| |_| .__/\__|
                                 |_|

FILENAME: newton.js
AUTHOR: Zachary Krepelka
DATE: Sunday, September 10th, 2023
ABOUT: a project for the exploration of programming languages
ORIGIN: https://github.com/zachary-krepelka/ascii-art-fractals.git
UPDATED: Sunday, October 26th, 2025 at 11:35 AM

*/

function fractal(px, max, tol, poly, colors, roots) {
	const der = diff(poly);
	const plane = makePlane(px);
	const unresolved = colors[colors.length - 1];
	for (let i = 0; i < px; i++) {
		for (let j = 0; j < px; j++) {
			let converged = false;
			let z = plane[i][j]; LOOP:
			for (let k = 0; k < max; k++) {
				for (let l = 0; l < roots.length; l++) {
					if (dist(z, roots[l]) <= tol) {
						converged = true;
						process.stdout.write(colors[l]);
						break LOOP;
					} // if
				} // for
				z = newton(poly, der, z);
			} // for
			if (!converged)
				process.stdout.write(unresolved);
		} // for
		process.stdout.write("\n");
	} // for
} // func

function makePlane(px) {
	let re = 0, im = 0;
	const stride = 2 / (px - 1);
	const plane = array3D(px, px, 2);
	for (let i = 0; i < px; i++) {
		for (let j = 0; j < px; j++) {
			if (j == 0) {
				re = -1; im = 1;
			} else {
				re += stride;
				im -= stride;
			} // if
			plane[i][j][0] = re;
			plane[j][i][1] = im;
		} // for
	} // for
	return plane;
} // func

function array3D(x, y, z) {
	const arr = [];
	for (let i = 0; i < x; i++) {
		arr.push([]);
		for (let j = 0; j < y; j++) {
			arr[i].push([]);
			for (let k = 0; k < z; k++) {
				arr[i][j][k] = null;
			} // for
		} // for
	} // for
	return arr;
} // func

newton = (poly, der, guess) =>
	sub(guess, div(eval(poly, guess), eval(der, guess)));

eval = (poly, z) =>
	poly.map((coeff, exp) => scl(coeff, pow(z, exp)))
	.reduce((sum, term) => add(sum, term));

diff = poly => poly.map((coeff, exp) => coeff * exp).slice(1);

dot = (z, w) => z[0] * w[0] + z[1] * w[1];

scl = (n, z) => [n * z[0], n * z[1]];

add = (z, w) => [z[0] + w[0], z[1] + w[1]];

sub = (z, w) => add(z, scl(-1, w));

div = (z, w) => {
	const re = z[0] * w[0] + z[1] * w[1];
	const im = z[1] * w[0] - z[0] * w[1];
	return scl(1/dot(w, w), [re, im]);};

abs = z => Math.sqrt(dot(z, z));

dist = (z, w) => abs(sub(z, w));

cis = arg => [Math.cos(arg), Math.sin(arg)];

arg = z => Math.atan2(z[1], z[0]);

pow = (z, n) => scl(abs(z) ** n, cis(n * arg(z)));

fractal(
	25, 256, 0.001,
	[-1, 0, 0, 1],
	[ // uses ANSI escape codes
		"\033[41m  \033[0m", // red
		"\033[42m  \033[0m", // green
		"\033[44m  \033[0m", // blue
		"\033[40m  \033[0m"  // black to signify the void
	],
	[
		[ 1  ,  0                 ],
		[-0.5,  0.8660254037844386],
		[-0.5, -0.8660254037844386]
	]
);

File: newton/newton.lua
--[[

 _
| |  _  _ __ _
| |_| || / _` |
|____\_,_\__,_|

FILENAME: newton.lua
AUTHOR: Zachary Krepelka
DATE: Tuesday, September 26th, 2023
ABOUT: a project for the exploration of programming languages
ORIGIN: https://github.com/zachary-krepelka/ascii-art-fractals.git
UPDATED: Sunday, October 26th, 2025 at 11:51 AM

--]]

function fractal(px, max, tol, poly, colors, roots)
	der = diff(poly)
	unresolved = colors[#colors]
	for _ = 1, px do
		for _ = 1, px do
			local _, z = coroutine.resume(nextComplex, px)
			converged = false
			for k = 1, max do
				for l = 1, #roots do
					if z:dist(roots[l]) <= tol then
						io.write(colors[l])
						converged = true
						break
					end
				end
				if converged then break end
				z = z:newton(poly, der)
			end
			if not converged then
				io.write(unresolved)
			end
		end
		print()
	end
end

function array3D(x, y, z)
	array = {}
	for i = 0, x - 1 do
		array[i] = {}
		for j = 0, y - 1 do
			array[i][j] = {}
			for k = 0, z - 1 do
				array[i][j][k] = 0
					end end end
						return array
							end

function makePlane(px)
	stride = 2 / (px - 1)
	plane = array3D(px, px, 2)
	for i = 0, px - 1 do
		for j = 0, px - 1 do
			if j == 0 then
				re, im = -1, 1
			else
				re = re + stride
				im = im - stride
			end
			plane[i][j][0] = re
			plane[j][i][1] = im
		end
	end
	return plane
end

nextComplex = coroutine.create(
	function (px)
		plane = makePlane(px)
		for i = 0, px - 1 do
			for j = 0, px - 1 do
				coroutine.yield(
					Complex:new(
						plane[i][j][0],
						plane[i][j][1]
					))
			end
		end
	end)

-- function printArray3D(arr)
-- 	print("[")
-- 	for i = 0, #arr do
-- 		io.write("  [ ")
-- 		for j = 0, #arr[i] do
-- 			io.write("[ ")
-- 			for k = 0, #arr[i][j] do
-- 				delimiter = k == #arr[i][j] and " " or ", "
-- 				io.write(arr[i][j][k], delimiter)
-- 			end
-- 			delimiter = j == #arr[i] and " " or ", "
-- 			io.write("]", delimiter)
-- 		end
-- 		delimiter = i == #arr and " " or ", "
-- 		io.write("]", delimiter, "\n")
-- 	end
-- 	print("]")
-- end

Complex = {re = 0, im = 1}

function Complex:new(real, imaginary)
	local z = {}
	setmetatable(z, self)
	self.__index = self
	z.re = real or 0
	z.im = imaginary or 0
	return z
end

function Complex:dot(z)
	return self.re * z.re + self.im * z.im
end

function Complex:scale(n)
	return Complex:new(n * self.re, n * self.im)
end

function Complex:__unm()
	return Complex:new(-self.re, -self.im)
end

function Complex:__add(z)
	return Complex:new(self.re + z.re, self.im + z.im)
end

function Complex:__sub(z)
	return self + -z
end


function Complex:__div(z)
	re = self.re * z.re + self.im * z.im
	im = self.im * z.re - self.re * z.im
	return (Complex:new(re, im)):scale(1/z:dot(z))
end

function Complex:abs()
	return math.sqrt(self:dot(self))
end

function Complex:dist(z)
	return (self - z):abs()
end

function Complex:__tostring()
	return self.re .. " + " .. self.im .. "i"
end

function Complex:cis(arg)
	return Complex:new(math.cos(arg), math.sin(arg))
end

function Complex:arg()
	return math.atan2(self.im, self.re)
end

function Complex:pow(n)
	return Complex:cis(n * self:arg()):scale(self:abs()^n)
end

function Complex:eval(poly)
	local sum = Complex:new(0,0)
	for i = 0, #poly - 1 do
		sum = sum + self:pow(i):scale(poly[i + 1])
	end
	return sum
end

function diff(poly)
	local der = {}
	for i = 1, #poly - 1 do
		table.insert(der, poly[i + 1] * i)
	end
	return der
end

function Complex:newton(poly, der)
	return self - self:eval(poly) / self:eval(der)
end

colors = {
	"\x1B[41m  \x1B[0m", -- red
	"\x1B[42m  \x1B[0m", -- green
	"\x1B[44m  \x1B[0m", -- blue
	"\x1B[40m  \x1B[0m"  -- black to signify the void
}

roots = {
	Complex:new( 1  ,  0                 ),
	Complex:new(-0.5,  0.8660254037844386),
	Complex:new(-0.5, -0.8660254037844386)
}

fractal(25, 256, 0.001, {-1, 0, 0, 1}, colors, roots)
